src/Ledger/Certs/Haskell.agda

```diff
@@ -0,0 +1,217 @@
+{-# OPTIONS --safe #-}
+
+open import Ledger.Prelude
+open import Ledger.Types.GovStructure
+
+module Ledger.Certs.Haskell
+  (gs : _) (open GovStructure gs)
+  where
+
+open import Ledger.Certs gs hiding ( DCert
+                                   ; CertEnv
+                                   ; DState
+                                   ; GState
+                                   ; CertState
+                                   ; GovCertEnv
+                                   ; _⊢_⇀⦇_,DELEG⦈_
+                                   ; _⊢_⇀⦇_,POOL⦈_
+                                   ; _⊢_⇀⦇_,GOVCERT⦈_
+                                   ; _⊢_⇀⦇_,CERTBASE⦈_
+                                   ; _⊢_⇀⦇_,CERT⦈_
+                                   ; _⊢_⇀⦇_,CERTS⦈_
+                                   ) public
+
+open import Tactic.Derive.DecEq
+
+open import Ledger.GovernanceActions gs
+open RwdAddr
+
+
+data DCert : Type where
+  delegate    : Credential → Maybe VDeleg → Maybe KeyHash → Coin → DCert
+  dereg       : Credential → Coin → DCert
+  regpool     : KeyHash → PoolParams → DCert
+  retirepool  : KeyHash → Epoch → DCert
+  regdrep     : Credential → Coin → Anchor → DCert
+  deregdrep   : Credential → DCert
+  ccreghot    : Credential → Maybe Credential → DCert
+
+record CertEnv : Type where
+  constructor ⟦_,_,_,_,_⟧ᶜ
+  field
+    epoch     : Epoch
+    pp        : PParams
+    votes     : List GovVote
+    wdrls     : RwdAddr ⇀ Coin
+    deposits  : Deposits
```

Comment:
Without `deposits` field in `CertEnv`, I haven't got the computational instances proofs to work out.  I was getting the deposits field of the `DCerts` constructor from the deposits variable in the `CertEnv` constructor... but I'll work on it.  I suspect I can work in out.  Will let you know if I get stuck.

Reply:
@WhatisRT does the `DelegEnv` still need a `deposits` field?

Reply:
These two cases fail if I remove the deposits field of `CertEnv`.

```agda
  Computational-CERT .completeness ⟦ _ , pp , _ , wdrls ⟧ᶜ ⟦ stᵈ , stᵖ , stᵍ ⟧ᶜˢ
    dCert@(delegate c mv mc d) ⟦ stᵈ' , stᵖ , stᵍ ⟧ᶜˢ (CERT-deleg h)
    with computeProof ⟦ pp , PState.pools stᵖ , _ ⟧ᵈᵉ stᵈ dCert | completeness _ _ _ _ h
  ... | success _ | refl = refl
  Computational-CERT .completeness ⟦ _ , pp , _ , wdrls ⟧ᶜ ⟦ stᵈ , stᵖ , stᵍ ⟧ᶜˢ
    dCert@(dereg c _) ⟦ stᵈ' , stᵖ , stᵍ ⟧ᶜˢ (CERT-deleg h)
    with computeProof ⟦ pp , PState.pools stᵖ , _ ⟧ᵈᵉ stᵈ dCert | completeness _ _ _ _ h
  ... | success _ | refl = refl
```

Here is the type-checking error message:

`./src/Ledger/Certs/Haskell/Properties.agda:89,21-25`
I'm not sure if there should be a case for the constructor refl,
because I get stuck when trying to solve the following unification
problems (inferred index ≟ expected index):
```agda
  (Functor-ComputationResult Functor.<$> (λ r → proj₁ r))
  ((λ { (yes p) → success (-, DELEG-delegate p)
      ; (no ¬p)
          → failure
            ((λ { (inj₁ x)
                    → "¬\n((z\n  : (_16 gs pp pools envDeps vDelegs sDelegs rwds dep c mv mc d\n     Interface.IsSet.IsSet.∉ c)\n    (Interface.IsSet.dom (.Axiom.Set.Theoryᵈ.th List-Modelᵈ) rwds)) →\n d ≡ .Ledger.PParams.PParams.keyDeposit pp)"
                ; (inj₂ x)
                    → (λ { (inj₁ x₁)
                             → "¬\n((z\n  : (_29 gs pp pools envDeps vDelegs sDelegs rwds dep c mv mc d\n     Interface.IsSet.IsSet.∈ c)\n    (Interface.IsSet.dom (.Axiom.Set.Theoryᵈ.th List-Modelᵈ) rwds)) →\n d ≡ 0)"
                         ; (inj₂ x₁)
                             → "¬\n(_39 gs pp pools envDeps vDelegs sDelegs rwds dep c mv mc d\n Interface.IsSet.IsSet.∈ mc)\n((.Axiom.Set.Theoryᵈ.th List-Modelᵈ Axiom.Set.Theory.∪\n  Axiom.Set.Theory.map (.Axiom.Set.Theoryᵈ.th List-Modelᵈ) just\n  (Interface.IsSet.dom (.Axiom.Set.Theoryᵈ.th List-Modelᵈ) pools))\n (Interface.HasSingleton.HasSingleton.❴\n  _51 gs pp pools envDeps vDelegs sDelegs rwds dep c mv mc d ❵\n  nothing))"
                         })
                      (dec-de-morgan x)
                })
             (dec-de-morgan ¬p))
      })
   ((((c ∈?
       proj₁ (replacement (λ r → proj₁ r) (proj₁ (DState.rewards stᵈ))))
      →-dec no (λ ()))
     →-dec
     map′
     (Data.Nat.Properties.≡ᵇ⇒≡ d
      (pp .Ledger.PParams.PParams.keyDeposit))
     (Data.Nat.Properties.≡⇒≡ᵇ d
      (pp .Ledger.PParams.PParams.keyDeposit))
     ((d Agda.Builtin.Nat.== pp .Ledger.PParams.PParams.keyDeposit)
      because
      T-reflects
      (d Agda.Builtin.Nat.== pp .Ledger.PParams.PParams.keyDeposit)))
    ×-dec
    ((c ∈?
      proj₁ (replacement (λ r → proj₁ r) (proj₁ (DState.rewards stᵈ))))
     →-dec
     map′ (Data.Nat.Properties.≡ᵇ⇒≡ d 0) (Data.Nat.Properties.≡⇒≡ᵇ d 0)
     ((d Agda.Builtin.Nat.== 0) because
      T-reflects (d Agda.Builtin.Nat.== 0)))
    ×-dec
    (mc ∈?
     proj₁
     (unions
      (proj₁
       (listing
        (proj₁
         (replacement just
          (proj₁ (replacement (λ r → proj₁ r) (proj₁ (PState.pools stᵖ)))))
         ∷ proj₁ (listing (nothing ∷ [])) ∷ []))))))) ≟ success stᵈ'
when checking that the pattern refl has type
(Functor-ComputationResult Functor.<$> (λ r → proj₁ r))
((λ { (yes p) → success (-, DELEG-delegate p)
    ; (no ¬p)
        → failure
          ((λ { (inj₁ x)
                  → "¬\n((z\n  : (_16 gs pp pools envDeps vDelegs sDelegs rwds dep c mv mc d\n     Interface.IsSet.IsSet.∉ c)\n    (Interface.IsSet.dom (.Axiom.Set.Theoryᵈ.th List-Modelᵈ) rwds)) →\n d ≡ .Ledger.PParams.PParams.keyDeposit pp)"
              ; (inj₂ x)
                  → (λ { (inj₁ x₁)
                           → "¬\n((z\n  : (_29 gs pp pools envDeps vDelegs sDelegs rwds dep c mv mc d\n     Interface.IsSet.IsSet.∈ c)\n    (Interface.IsSet.dom (.Axiom.Set.Theoryᵈ.th List-Modelᵈ) rwds)) →\n d ≡ 0)"
                       ; (inj₂ x₁)
                           → "¬\n(_39 gs pp pools envDeps vDelegs sDelegs rwds dep c mv mc d\n Interface.IsSet.IsSet.∈ mc)\n((.Axiom.Set.Theoryᵈ.th List-Modelᵈ Axiom.Set.Theory.∪\n  Axiom.Set.Theory.map (.Axiom.Set.Theoryᵈ.th List-Modelᵈ) just\n  (Interface.IsSet.dom (.Axiom.Set.Theoryᵈ.th List-Modelᵈ) pools))\n (Interface.HasSingleton.HasSingleton.❴\n  _51 gs pp pools envDeps vDelegs sDelegs rwds dep c mv mc d ❵\n  nothing))"
                       })
                    (dec-de-morgan x)
              })
           (dec-de-morgan ¬p))
    })
 ((((c ∈?
     proj₁ (replacement (λ r → proj₁ r) (proj₁ (DState.rewards stᵈ))))
    →-dec no (λ ()))
   →-dec
   map′
   (Data.Nat.Properties.≡ᵇ⇒≡ d
    (pp .Ledger.PParams.PParams.keyDeposit))
   (Data.Nat.Properties.≡⇒≡ᵇ d
    (pp .Ledger.PParams.PParams.keyDeposit))
   ((d Agda.Builtin.Nat.== pp .Ledger.PParams.PParams.keyDeposit)
    because
    T-reflects
    (d Agda.Builtin.Nat.== pp .Ledger.PParams.PParams.keyDeposit)))
  ×-dec
  ((c ∈?
    proj₁ (replacement (λ r → proj₁ r) (proj₁ (DState.rewards stᵈ))))
   →-dec
   map′ (Data.Nat.Properties.≡ᵇ⇒≡ d 0) (Data.Nat.Properties.≡⇒≡ᵇ d 0)
   ((d Agda.Builtin.Nat.== 0) because
    T-reflects (d Agda.Builtin.Nat.== 0)))
  ×-dec
  (mc ∈?
   proj₁
   (unions
    (proj₁
     (listing
      (proj₁
       (replacement just
        (proj₁ (replacement (λ r → proj₁ r) (proj₁ (PState.pools stᵖ)))))
       ∷ proj₁ (listing (nothing ∷ [])) ∷ [])))))))
≡ success stᵈ'
```

Reply:
I didn't see that when I looked at it during the review, but you'll also have to remove the deposits from `DelegEnv`. Then it should work just fine (I tried it and needed no adjustments to `Computational` from your branch).